src/Func.cc

```diff
@@ -506,31 +515,77 @@ void ScriptFunc::CreateCaptures(Frame* f)
 	if ( ! captures )
 		return;
 
-	// Create a private Frame to hold the values of captured variables,
-	// and a mapping from those variables to their offsets in the Frame.
-	delete captures_frame;
-	delete captures_offset_mapping;
-	captures_frame = new Frame(captures->size(), this, nullptr);
-	captures_offset_mapping = new OffsetMap;
+	// Create *either* a private Frame to hold the values of captured
+	// variables, and a mapping from those variables to their offsets
+	// in the Frame; *or* a ZVal frame if this script has a ZAM-compiled
+	// body.
+	ASSERT(bodies.size() == 1);
+
+	if ( bodies[0].stmts->Tag() == STMT_ZAM )
+		captures_vec = std::make_unique<std::vector<ZVal>>();
+	else
+		{
+		delete captures_frame;
+		delete captures_offset_mapping;
+		captures_frame = new Frame(captures->size(), this, nullptr);
+		captures_offset_mapping = new OffsetMap;
+		}
 
 	int offset = 0;
 	for ( const auto& c : *captures )
 		{
-		auto v = f->GetElementByID(c.id);
+		auto v = f->GetElementByID(c.Id());
 
 		if ( v )
 			{
-			if ( c.deep_copy || ! v->Modifiable() )
+			if ( c.IsDeepCopy() || ! v->Modifiable() )
 				v = v->Clone();
 
-			captures_frame->SetElement(offset, std::move(v));
+			if ( captures_vec )
+				// Don't use v->GetType() here, as that might
+				// be "any" and we need to convert.
+				captures_vec->push_back(ZVal(v, c.Id()->GetType()));
+			else
+				captures_frame->SetElement(offset, std::move(v));
 			}
 
-		(*captures_offset_mapping)[c.id->Name()] = offset;
+		else if ( captures_vec )
+			captures_vec->push_back(ZVal());
+
+		if ( ! captures_vec )
+			(*captures_offset_mapping)[c.Id()->Name()] = offset;
```

Comment:
Personally, I like the readability of using an `=` assignment rather than a non-visually-clear-it's-an-assignment function call ... but not a biggie, so will tweak with the next push.

Reply:
Now pushed.